test_util/test_runner.py

```diff
@@ -34,71 +58,57 @@ def proxy_registry_as_local(tunnel, agent_list, test_dir):
             '-oExitOnForwardFailure=yes',
             '-R', '5000:127.0.0.1:5000']
     p_list = []
-    log.info('Starting registry proxy')
-    for agent in agent_list:
-        agent_cmd_list = cmd_list + [agent]
-        p = Process(target=tunnel.remote_cmd, args=(agent_cmd_list,))
+    log.info('Starting proxy of port 5000 from {} to {}'.format(host_list, tunnel.host))
+    for host in host_list:
+        host_cmd_list = cmd_list + [host]
+        p = Process(target=tunnel.remote_cmd, args=(host_cmd_list,))
         p.start()
         p_list.append(p)
     yield
     for p in p_list:
         p.terminate()
 
 
-def pkg_filename(relative_path):
-    return pkg_resources.resource_filename(__name__, relative_path)
-
-
-def setup_integration_test(tunnel, test_dir):
-    """Transfer resources and issues commands on host to build test app,
-    host it on a docker registry, and prepare the integration_test container
-    Note: we perform configuration of the nodes via a single establish tunnel
-        so that the test may run through a load balance (i.e. Azure tests)
+def prepare_test_registry(tunnel, test_dir):
+    """integration_test.py requires a registry loaded with an based on test_server.py
+    The registry only runs on one host, but each host will have 127.0.0.1 forwarded to
+    a remote registry (to bypass security) using remote_port_forwarding()
+    1. Build test_server app locally
+    2. Export test_server to tunnel host
+    3. load into docker
+    4. Push test_server to registry
 
     Args:
-        test_dir (str): path to be used for setup and file transfer on host
-
-    Returns:
-        result from async chain that can be checked later for success
+        tunnel: SSHTunnel instance
+        test_dir (str): path to be used for landing docker archives
     """
+    tunnel.remote_cmd(['docker', 'run', '-d', '-p', '5000:5000', '--restart=always', '--name',
+                       'test_registry', 'registry:2'])
+    test_server_image = '127.0.0.1:5000/test_server'
+    log.info('Building and exporting '+test_server_image)
+    temp_dir = tempfile.mkdtemp()
     test_server_docker = pkg_filename('docker/test_server/Dockerfile')
     test_server_script = pkg_filename('docker/test_server/test_server.py')
-    pytest_docker = pkg_filename('docker/py.test/Dockerfile')
-    log.info('Setting up integration_test.py to run on ' + tunnel.host)
+    shutil.copy(test_server_docker, join(temp_dir, 'Dockerfile'))
+    shutil.copy(test_server_script, join(temp_dir, 'test_server.py'))
+    check_call(['docker', 'build', '-t', test_server_image, '.'], cwd=temp_dir)
+    check_call(['docker', 'save', '-o', join(temp_dir, 'test_server.tar'), test_server_image])
```

Comment:
I'm not a huge fan of keeping the registry on the teamcity host and tunneling to the test host which then maintains a tunnel to all the dcos nodes.

The reason that I changed this from doing your favored approach was that I was hoping to utilize the fact that the test_server and registry layers have likely already been built and cached on the test runner node, whereas if we build on remote we will definitely have to pull in all the deps.

Also, just for kicks and giggles I thought it would be cool if this test was made in such a way that none of the test nodes needed a connection outside the local network. E.G. if you wanted to use this as a CLI tool to trigger tests against arbitrary clusters, it would be nice if your test node could be in a restricted network

So, that was the logic, but those features aren't currently being asked for, so I can definitely revert this if desired